quinn-proto/src/connection.rs

```diff
@@ -1964,47 +1976,44 @@ impl Connection {
         &mut self,
         handshake: bool,
         packet: &mut Packet,
+        crypto_update: Option<Crypto>,
     ) -> Result<Option<u64>, Option<TransportError>> {
         if packet.header.is_retry() {
             // Retry packets are not encrypted and have no packet number
             return Ok(None);
         }
-        let (key_phase, number) = match packet.header {
-            Header::Short {
-                key_phase, number, ..
-            } if !handshake => (key_phase, number),
-            Header::Initial { number, .. } | Header::Long { number, .. } if handshake => {
-                (false, number)
-            }
+        let number = match packet.header {
+            Header::Short { number, .. } if !handshake => number,
+            Header::Initial { number, .. } | Header::Long { number, .. } if handshake => number,
             _ => {
                 return Err(None);
             }
         };
         let number = number.expand(self.rx_packet + 1);
-        if key_phase != self.key_phase {
-            if number <= self.rx_packet {
-                // Illegal key update
-                return Err(Some(TransportError::PROTOCOL_VIOLATION));
-            }
-            let new = self.crypto.as_mut().unwrap().update();
-            new.decrypt(number, &packet.header_data, &mut packet.payload)
-                .map_err(|()| None)?;
 
-            let old = mem::replace(self.crypto.as_mut().unwrap(), new);
+        if crypto_update.is_some() && number <= self.rx_packet {
+            warn!(self.log, "recieved an illegal key update");
+            return Err(Some(TransportError::PROTOCOL_VIOLATION));
+        }
```

Comment:
Sorry, one more thing: can you add an `else if` here that throws another `PROTOCOL_VIOLATION` if `handshake == true && crypto_update.is_some()`?

Reply:
Yes, sure, good catch !